### DISCUSSION TITLE
serialtest.py Script Error

### DISCUSSION THREAD

**Discussion by deh2025** [jgyates/genmon#1056] [Questions]
I suspect that I doing something rather stupid, but I cannot run the serialtest.py while attempting to bench test the Pi (3b+) in conjunction with a PintSizeMe OPTO-ISOLATED HAT and Molex cable prior to installation later this week.  Pins 7*8 are looped for tx/rx testing.

The error received is - XXXX@GeneracGenmon:~ $ serialtest.py
bash: serialtest.py: command not found

The Monitor is running using two-factor authentication and is displaying CPU temperature and WiFi strength.  Email tests were also successful.  So, I know that the GENMON script is running correctly.

Thanks!

**Comment by skipfire** [comment]
First you will have to change to the folder where serialtest is, `cd genmon/OtherApps` I believe and then try running it.  Second, it will still fail with the opto-isolated HAT as you won't have power and ground on the Molex plug so the opto-isolator will be unpowered on that side.  When bench testing prior to packing them for sale I have a special test harness that supports adding 12v or 5v power specifically so I can test the opto-isolated products.  Third, genmon has to be not running when using the serialtest.py script, otherwise the serial port is already in use and the serialtest.py script cannot access it.

**Comment by jgyates** [comment]
This page has details on testing the serial port:

https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting

The "command not found" error is the operating system telling you that it can not find the command you are typing. There are possibly two things wrong.

1) you are not in the ~./genmon/OtherApps folder (note the tilde character '~' represents your home folder associated with your user name)
2) serialtest.py is a python program so you must run it with the python3 program.

once you log in via ssh you should type:

     cd genmon
     cd OtherApps
     python3 serialtest.py

If you get an error with the above about a managed environment, then you are likly running  latest version of the pi OS (Bookworm version 12). If that is the case this page offers detail about running in a managed python environment:

https://github.com/jgyates/genmon/wiki/Appendix-S---Working-in-a-Managed-Python-Environment
In a managed python environment these are the commands to run:

     cd genmon
     source ./genenv/bin/activate
      sudo ./genenv/bin/python3 ./OtherApps/serialtest.py 
      deactivate

**Reply by jgyates** [reply]
to stop genmon, from the genmon folder type

      ./startgenmon.sh stop

**Comment by deh2025** [comment]
Thank you both for your replies!  I removed the HAT and hardwired GPIO 14 to GPIO 15. I must still be doing something incorrect.  Below was the outcome.

~ $ cd genmon
~/genmon $ ./startgenmon.sh stop
Stopping genmon python scripts
Stopping....
~/genmon $ cd OtherApps
~/genmon/OtherApps $ python3 serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5b0
error opening serial port: [Errno 13] could not open port /dev/serial0: [Errno 13] Permission denied: '/dev/serial0'

Try stopping genmon.

Error opening Serial Port /dev/serial0

**Comment by jgyates** [comment]
I will need some more info. 

Did you install the software using this procedure: https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software

or did you purchase a preinstalled unit?

What is the output of this command:

      ls -la /dev/ser*

What OS variant did you install? Raspbian, 32 bit or 64 bit?

You also have a beta version of the serial python library (V3.5b). This happens when the install pulls the pyserial library from an outdated mirror. Type this to correct:

     sudo apt-get remove python3-serial
     sudo pip3 uninstall pyserial
     sudo pip3 install pyserial

Note one of the first two may fail. Then run serialtest again. You can show the version of the python libraries installed with this command:

      sudo pip3 freeze

If none of this works then send a pic of your loopback.

**Comment by deh2025** [comment]
Hi,
Yes, I only installed the software using your github procedure on a new 64GB Micro SDXC.

Command output is -
~ $   ls -la /dev/ser*
lrwxrwxrwx 1 root root 7 Feb 18 18:13 /dev/serial0 -> ttyAMA0
lrwxrwxrwx 1 root root 5 Feb 18 18:13 /dev/serial1 -> ttyS0

The OS variant installed is Raspbian, 32 bit.

I am unsure whether the script below addressed the beta version download.  It appears that it was removed but replaced again with 3.5.

INPUT
sudo apt-get remove python3-serial
 sudo pip3 uninstall pyserial
 sudo pip3 install pyserial

OUTPUT
Successfully uninstalled pyserial-3.5
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting pyserial
  Using cached https://www.piwheels.org/simple/pyserial/pyserial-3.5-py2.py3-none-any.whl (90 kB)
Installing collected packages: pyserial
Successfully installed pyserial-3.5


~ $ cd genmon
~/genmon $ ./startgenmon.sh stop
Stopping genmon python scripts
Stopping....
~/genmon $ cd OtherApps
~/genmon/OtherApps $ python3 serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
error opening serial port: [Errno 13] could not open port /dev/serial0: [Errno 13] Permission denied: '/dev/serial0'

Try stopping genmon.

Error opening Serial Port /dev/serial0

**Comment by jgyates** [comment]
That fixed the issue with the beta version of the serial library.

what is the output of this command:

      fuser /dev/serial0

If the output is nothing then this means the port is not open by anyone else. I am guessing this will be the case as the error that serailtest.py normally outputs is a little different if another process is using the port. it would look like this:
<img width="566" alt="Screenshot 2024-02-18 at 8 52 49 PM" src="https://github.com/jgyates/genmon/assets/4513286/73872e42-ce1b-4bb3-ba80-a7cbd84224b7">

If the output shows a number that is the process number. Fill in the NUMBER below with the number shown in the output above using this command:

  ps -p NUMBER -o comm=

what is this output?

**Comment by deh2025** [comment]
I tried fuser /dev/serial0 in all directories with no output.  So your guess was absolutely correct!

**Comment by jgyates** [comment]
what is odd is that you are getting a permission error, but the permissions on the file show that the serial port has permissions for all users. 

How did you create the user you are using to log into the pi to run these test?

Try this test

     sudo python3 serialtest.py

This will run the test with administrator privileges

**Comment by deh2025** [comment]
This is the output.  Should I have run  ./startgenmon.sh stop prior to doing so?

~/genmon/OtherApps $ sudo python3 serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
Serial port opened
write data: sent test string
waiting to received data....
FAILED: Sent data does not match receive. Received 0 bytes

**Comment by jgyates** [comment]
yes, try running this again without genmon running and without sudo

**Comment by jgyates** [comment]
genmon cannot run at the same time as serial test. Only one application can have access to the serial port.

**Comment by deh2025** [comment]
I wish that I had better news.

~/genmon $ ./startgenmon.sh stop
Stopping genmon python scripts
Stopping....
~/genmon $ cd OtherApps
~/genmon/OtherApps $ python3 serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
error opening serial port: [Errno 13] could not open port /dev/serial0: [Errno 13] Permission denied: '/dev/serial0'

Try stopping genmon.

Error opening Serial Port /dev/serial0

**Comment by deh2025** [comment]
This was a little different using, sudo python3 serialtest.py.  This time it received 6 bytes.

~/genmon/OtherApps $ sudo python3 serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
Serial port opened
write data: sent test string
waiting to received data....
FAILED: Sent data does not match receive. Received 6 bytes

**Comment by deh2025** [comment]
SUCCESS!!
~/genmon/OtherApps $ sudo python3 serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
Serial port opened
write data: sent test string
waiting to received data....
PASSED! Loopback successful

**Comment by jgyates** [comment]
Great, what changed?

**Comment by deh2025** [comment]
Not sure, but it passed twice.

**Comment by jgyates** [comment]
ok. I am going to assume intermittent connection on your serial loopback.

**Comment by deh2025** [comment]
Let me know when your PayPal donation link is running, and I will gladly make a donation.  ;)  I tried a couple times earlier in the week, but it timed out.  I really appreciate your dedication to this effort!

**Comment by jgyates** [comment]
No problem, I am glad it is working. I just checked the link, it is working for me: https://www.paypal.com/paypalme/jgyates

**Comment by jgyates** [comment]
Thanks!

**Comment by deh2025** [comment]
You see something coming your way.  Thank you!
